### PR TITLE
msp/operationdocs: write to Notion instead of Markdown

### PR DESCRIPTION
**Pull request by bobheadxi**
Closes https://github.com/sourcegraph/managed-services/issues/1076 aka closes CORE-28 - https://github.com/sourcegraph/managed-services/pull/1332 also automated the updates.

Notes:

1. Notion anchors are ID-based, so we strip all anchor links because we cannot generate them in one pass. `notionreposync` may implement a mechanism for this in the future (filed https://github.com/sourcegraph/notionreposync/issues/8), but for now we don't have a great way around this, especially because of the next point
2. In-place updates are hard because of the block structure, so we destroy page contents and recreate the page every time. This causes a "flicker" as a viewer may see the page disappear slowly before their eyes (we can only delete things 1 block at a time). `notionreposync` may may implement an improved mechanism for this in the future (filed https://github.com/sourcegraph/notionreposync/issues/7)
3. There's something funky going wrong with line breaks, filed https://github.com/sourcegraph/notionreposync/issues/9 - it hurts readability but it's not unmanageable.
4. Sadly Notion does not allow API file uploads (😡 https://developers.notion.com/docs/working-with-files-and-media#uploading-files-and-media-via-the-notion-api), so we generate them into the managed-services repo (https://github.com/sourcegraph/managed-services/pull/1332) and then just link to those diagrams from the generated page. We use a Markdown file that renders the SVG because the native SVG viewer sucks.
5. Made misc changes to operationdocs output where Notion's version is noticeably worse, or difficult to support (tables, lists-in-admonitions, etc)

Depends on various improvements upstream in https://github.com/sourcegraph/notionreposync:

- https://github.com/sourcegraph/notionreposync/pull/4
- https://github.com/sourcegraph/notionreposync/pull/5
- https://github.com/sourcegraph/notionreposync/pull/6

Follow-up improvements:

- CORE-105
- CORE-106

## Test plan

```
sg msp ops generate-handbook-pages
```

image

https://sourcegraph.notion.site/Build-Tracker-infrastructure-operations-bd66bf25d65d41b4875874a6f4d350cc

image

https://github.com/sourcegraph/managed-services/pull/1332

### CODE REVIEW

**Review by jhchabran — APPROVED** [review]
@bobheadxi a `sg bazel configure godeps` was missing, pushed a fix.

File: dev/managedservicesplatform/operationdocs/index.go [hunk 1]
```diff
@@ -11,55 +11,40 @@ import (
...
 
-// HandbookDirectory designates where in sourcegraph/handbook operation docs should go.
-//
-// Place under top-level 'engineering/managed-services' since it's too much work
-// to find the appropriate team-specific content tree, and they change frequently.
-const HandbookDirectory = "content/departments/engineering/managed-services"
+// IndexNotionPageID designates where in Notion the contents of
+// operationdocs.RenderIndexPage should go.
+// https://www.notion.so/sourcegraph/Managed-Services-0d0b709881674eee9dca4202de9f93b1
+func IndexNotionPageID() string { return "0d0b709881674eee9dca4202de9f93b1" }
```
[the remark this answers is not in the document]
Comment (bobheadxi) [reply dev/managedservicesplatform/operationdocs/index.go:17]: This is pretty Sourcegraph-specific, I think we can aim to keep the library fairly generalized, WDYT?

[the remark this answers is not in the document]
Comment (jhchabran) [reply dev/managedservicesplatform/operationdocs/index.go:17]: I actually commented on the wrong thing, I was referring to the url builder helper: https://github.com/sourcegraph/sourcegraph/pull/62325/files#diff-825fa8584306019ac3f38d7af8540c07a0dc12ed85b1a5be338d65ea0218de9fR26

[the remark this answers is not in the document]
Comment (bobheadxi) [reply dev/managedservicesplatform/operationdocs/index.go:17]: Ah - I mean, that's also Sourcegraph-specific as the root URL is the Soucegraph public Notion site 😆

[the remark this answers is not in the document]
Comment (jhchabran) [reply dev/managedservicesplatform/operationdocs/index.go:17]: Yeah, ofc, I was thinking about `notionsync.New("sourcegraph").WebURL(id)` would get you `https://notion.so/sourcegraph/<id>`

File: dev/sg/msp/helpers.go [hunk 2]
```diff
@@ -303,38 +303,48 @@ func generateTerraform(service *spec.Spec, opts generateTerraformOptions) error
 
 		}
 
-		pending.Complete(output.Styledf(output.StyleSuccess,
-			"[%s] Category %q environment %q infrastructure assets generated!",
-			serviceID, env.Category, env.ID))
-	}
+		// We must persist diagrams somewhere for reference in Notion, since
+		// Notion does not allow us to upload files via API.
+		// https://developers.notion.com/docs/working-with-files-and-media#uploading-files-and-media-via-the-notion-api
```
Comment (unknwon) [inline dev/sg/msp/helpers.go:308]: CleanShot 2024-05-03 at 10 59 19@2x

Any chance we just tell Notion to embed our images (as externally hosted)?

**Review by unknwon — APPROVED** [review]

File: dev/sg/msp/helpers.go (hunk 2 above)
Comment (bobheadxi) [reply dev/sg/msp/helpers.go:308]: This mechanism works by simply giving it an external image URL. I thought about provisioning a bucket and placing the diagram SVG in there (got halfway to implementing it actually: https://github.com/sourcegraph/sourcegraph/pull/62370), but ultimately decided that this wasn't worth the time right now 😆

Comment (bobheadxi) [reply dev/sg/msp/helpers.go:308]: Embedding from `sourcegraph/managed-services` also doesn't work as that is a private repo, the image results in a 404

Comment (bobheadxi) [reply dev/sg/msp/helpers.go:308]: I filed [CORE-106](https://linear.app/sourcegraph/issue/CORE-106/msp-ops-host-service-diagrams-so-that-they-can-be-embedded-in-notion) as a follow-up

File: dev/managedservicesplatform/operationdocs/internal/markdown/headings/headings.go [hunk 3]
```diff
@@ -0,0 +1,27 @@
+package headings
+
+import "unicode"
+
+// SanitizeHeadingID returns a sanitized anchor name for the given text.
+// Taken from https://github.com/shurcooL/sanitized_anchor_name/blob/master/main.go#L14:1
+// copy from https://sourcegraph.com/github.com/gomarkdown/markdown@663e2500819c19ed2d3f4bf955931b16fa9adf63/-/blob/parser/block.go?L83-104
```
[the remark this answers is not in the document]
Comment (bobheadxi) [reply dev/managedservicesplatform/operationdocs/internal/markdown/headings/headings.go:7]: This is just a code move and some leftover change from a brief experiment to make anchor links work i.e. https://github.com/sourcegraph/notionreposync/issues/8 😅  I'll tidy up the docstrings though

**Pull request merged**